src/hlsBinaries.ts

```diff
@@ -308,14 +331,29 @@ async function getLatestSuitableHLS(
             : await getProjectGHCVersion(wrapper, workingDir, logger);
 
     // get installable HLS that supports the project GHC version (this might not be the most recent)
-    const latestMetadataHls =
-        projectGhc !== null ? await getLatestHLSforGHC(context, storagePath, projectGhc, logger) : null;
-    if (latestMetadataHls === null) {
-        const noMatchingHLS = `No HLS version was found for supporting GHC ${projectGhc}.`;
-        window.showErrorMessage(noMatchingHLS);
-        throw new Error(noMatchingHLS);
-    } else {
+    const latestMetadataHls = await getLatestHLSfromMetadata(context, storagePath, projectGhc, logger);
+    const latestGhcupHls = await getLatestHLSfromGHCup(context, storagePath, logger, projectGhc).then(e => e === null ? null : e[0]);
```

Comment:
Here's where we get a possibly self-compiled already installed HLS.

Reply:
Worthy not: we only consider the *latest* self-compiled...

Reply:
There is another edge case: if you compile HLS 1.6.1.0 manually instead of via bindist, the supported GHC versions differ from the metadata.

I'm not sure that's worth to support explicitly.

Reply:
> Worthy not: we only consider the latest self-compiled...

What does that mean? If I have compiled version 1.6.1.0, and then I compile version 1.5.0.0, does that mean version 1.5.0.0 is used (since it is the most recent compiled HLS version), or that no matter what, we only use 1.6.1.0 and if there is no version 1.6.1.0, we would not use the self compiled version 1.5.0.0 either? Does this take into account what the latest supported HLS version relative to a GHC version is? E.g. when 8.10.4 supports only up to 1.4.0.0, and I compile the version 1.5.0.0 for it (because that's possible in theory) would the extension pick that up correctly?

Reply:
If 1.6.1.0 is the latest properly released one and you have 1.7.0.0 and 1.8.0.0 compiled, the extension will only bother looking at 1.8.0.0.

There are a number of complicated interactions here, e.g. it's possible that you compiled 1.6.1.0 as well and it doesn't support the project GHC, because it's only built for one GHC (unlike the bindist).

A way out here would be this:

1. Map A: gather all ghcup-installed HLSes and their corresponding supported GHC version, regardless of the HLS metadata file
2. Map B: gather HLS metadata
3. merge both this way:
    - if a HLS version exists both in A and B, only consider GHC versions from A (because `ghcup install hls <ver>` would not overwrite the already existing one... and the existing one might have *less* supported GHC versions)
    - if a HLS version exists only in A, pick A
    - if a HLS version exists only in B, pick B

Reply:
I'm starting to think this functionality belongs into ghcup. It is non-trivial.

Reply:
It is non-trivial. Adding it to ghcup directly sounds like a nice idea!